### PR TITLE
Issue 15384 - assignment is sometimes still accepted as a condition

### PR DESCRIPTION
**Pull request by 9rnsr**
https://issues.dlang.org/show_bug.cgi?id=15384
1. Add CommaExp.toBoolean to show error message.
2. Add checkAssignmentAsCondition and call it before semantic analysis to provide informative diagnostic.

### CODE REVIEW

File: src/statement.d [hunk 1]
```diff
@@ -81,6 +81,19 @@ extern (C++) LabelStatement checkLabeledLoop(Scope* sc, Statement statement)
...
 }
 
+Expression checkAssignmentAsCondition(Expression e)
+{
+    Expression e0;
+    Expression ec = Expression.extractLast(e, &e0);
+    if (ec.op == TOKassign)
+    {
+        e.error("assignment cannot be used as a condition, perhaps == was meant?");
+        auto ae = cast(AssignExp)ec;
```
Comment (WalterBright) [inline src/statement.d]: Should return `ErrorExp` here, not try to repair things.

File: src/statement.d [hunk 2]
```diff
@@ -81,6 +81,19 @@ extern (C++) LabelStatement checkLabeledLoop(Scope* sc, Statement statement)
...
+        ec = new EqualExp(TOKequal, ec.loc, ae.e1, ae.e2);
+    }
+    return Expression.combine(e0, ec);
```
Comment (WalterBright) [inline src/statement.d]: Should just return `e` here.

File: src/statement.d [hunk 3]
```diff
@@ -81,6 +81,19 @@ extern (C++) LabelStatement checkLabeledLoop(Scope* sc, Statement statement)
     return null;
 }
 
+Expression checkAssignmentAsCondition(Expression e)
```
Comment (WalterBright) [inline src/statement.d]: Need Ddoc comment for function.

File: src/expression.d [hunk 4]
```diff
@@ -13976,7 +13985,10 @@ public:
 
     override Expression toBoolean(Scope* sc)
     {
-        e2 = e2.toBoolean(sc);
+        auto ex2 = e2.toBoolean(sc);
```
Comment (WalterBright) [inline src/expression.d]: How about simply:
```
e2 = e2.toBoolean(sc);
return e2.op == TOKerror ? e2 : this;
```
?

File: src/statement.d (hunk 1 above)
Comment (9rnsr) [inline src/statement.d]: This is for more informative error report. Read the code comment I've added more.

File: src/statement.d (hunk 2 above)
Comment (9rnsr) [inline src/statement.d]: Ditto for the error report.

File: src/statement.d (hunk 3 above)
Comment (9rnsr) [inline src/statement.d]: Done.

File: src/expression.d (hunk 4 above)
Comment (9rnsr) [inline src/expression.d]: `e2 = e2.toBoolean(sc);` sets `this.e2` to `ErrorExp`. It's not good during error gagging.

Comment (WalterBright) [inline src/expression.d]: The whole scheme of error reporting in DMD is to mark the trees as TOKerror, and not try and "repair" the trees, ever.

File: src/statement.d [hunk 5]
```diff
@@ -81,6 +81,33 @@ extern (C++) LabelStatement checkLabeledLoop(Scope* sc, Statement statement)
...
+    Expression ec = Expression.extractLast(e, &e0);
+    if (ec.op == TOKassign)
+    {
+        e.error("assignment cannot be used as a condition, perhaps == was meant?");
+
+        /* If just return ErrorExp here, no more errors will be reported
+         * even if either of ae operands contains any errors.
+         * By the rewriting, following semantic analysis will report the errors
+         * in ae.e1 and ae.e2.
+         */
```
Comment (WalterBright) [inline src/statement.d]: This is not how things must work. The point is not to get all the errors, which is impossible. The point is to mark the node as TOKerror, and mark any node dependent on a TOKerror as also TOKerror. Trying to "repair" the node destroys the efficacy of this.

The point of all this is to remove dependency on checking the global error count - should just check the top level node for TOKerror.

File: src/expression.d (hunk 4 above)
Comment (9rnsr) [inline src/expression.d]: Currently, if a semantic analysis is done in error gagging section, the analyzed original AST might be reused. Therefore, if a sub AST generates `ErrorExp`, it should not be stored back into the original AST. If not, the stored `ErrorExp` would appear in the whole semantic analysis result, and of course the result AST will become invalid.

Until we completely change the gagging mechanism, we should use such the idiomatic code.

File: src/statement.d (hunk 5 above)
Comment (9rnsr) [inline src/statement.d]: OK, make sense. Changed to just return `ErrorExp`.

**Comment by WalterBright** [comment]
Auto-merge toggled on

**Comment by ibuclaw** [comment]
This may have caused a regression.

**Pull request merged**